Introduce `categories_as_tabs` & group engines in tabs

not-my-profile opened the pull request:
## What does this PR do?

Previously all categories were displayed as tabs in the UI. This PR changes this so that only categories
listed under `categories_as_tabs` in `settings.yml` are displayed.

This lets us introduce more categories without cluttering up the user-interface.
These new non-tab categories are then:

* used to group the engine tables in the preferences
* used to group the engine tables in the documentation
* searchable with bangs e.g. `!repos youtube-dl`

For details please see the commit messages.

|documentation|preferences|
|--|--|
|![image](https://user-images.githubusercontent.com/73739153/148104586-c6572274-dc64-4912-8c3f-fde97ef8832c.png)| ![image](https://user-images.githubusercontent.com/73739153/148104741-5879088e-c859-451e-aef3-c527271157ce.png)

## Why is this change important?

The grouping provides a better overview and the bangs allow for new queries.

## Related issues
#623, #621

not-my-profile committed:
[enh] settings.yml: allow granular overwrites for about

not-my-profile committed:
[enh] introduce categories_as_tabs

Previously all categories were displayed as search engine tabs.
This commit changes that so that only the categories listed under
categories_as_tabs in settings.yml are displayed.

This lets us introduce more categories without cluttering up the UI.
Categories not displayed as tabs  can still be searched with !bangs.

not-my-profile committed:
[enh] add more categories

not-my-profile committed:
[enh] display language of engines in preferences

return42 commented on docs/admin/engines/settings.rst:
```diff
@@ -219,6 +219,25 @@ Communication with search engines.
+    - science
+    - files
+    - social media
+
```
> I documented `main_settings`

Can you please add a documentation of main- & secondary categories here, like you gave in the comments above https://github.com/searxng/searxng/pull/634#issuecomment-1001010436

I also miss a description for the users / https://github.com/searxng/searxng/pull/634#issuecomment-1001011772 ...

>> they can now search wikis with `!wikis`, repositories with `!repositories`, etc.

> One more, I think we do need a description for the users https://searxng.github.io/searxng/user/search_syntax.html

return42 commented on searx/templates/oscar/preferences.html:
```diff
@@ -348,7 +348,11 @@ <h3>{{ _('Engines') }}</h3>
+                        {% if not (group == 'powered by other sources' and loop.index == 1) %}
+                        <tr><th colspan=9>{{_(group)}}</th></tr>
+                        {% endif %}
+                        {% for search_engine in engines %}
```
> From where can the user get the informations what secondary categories are defined ... 

I think we need a list of secondary categories documented somewhere in the preferences .. not just grouping ...

> I now updated the preference templates to also group categories by the "second category".

I am sorry to say, but this sort order (by secondary category) confused me / tooked a while I understand ... and when I look for the setting of an engine (aka module) I miss the alphabetic order ... is it only me?

![grafik](https://user-images.githubusercontent.com/554536/147405018-93e15f91-7bb8-41fb-a47f-f86dcee4e847.png)

not-my-profile replied:
I think that's because the Simple theme aligns the new category headings in the center. It think it should look better when it's aligned to the left (and maybe has a different background for emphasis).

not-my-profile replied:
>I think we need a list of secondary categories documented somewhere in the preferences .. not just grouping ...

I don't think that's necessary.

I updated the simple theme, I think the subcategories now look much better:

![image](https://user-images.githubusercontent.com/73739153/147425581-f5f69fe6-8139-4871-87fd-8b5f2e0fdb0d.png)

return42 commented on searx/engines/google.py:
```diff
@@ -41,7 +41,7 @@
 
 # engine dependent config
-categories = ['general']
+categories = ['general', 'powered by google']
```
I understand categories like 'repos', 'wikis', 'database' ... theses categories classify engines ... but I do not understand how 'powered by ...' fits into this scheme of classifications.

We need a documentation of the intention of subcategories, we should name what _default_ set of subcategories SearXNG has and we should give some rules at hand how to classify new added engines well in the right subcategory.

return42 reviewed (commented):
Thanks for your additional work .. I left some comments.  Over all I think grouping by "subcateories" is a useful addition, but I really miss a better documentation of this *feature* and the concept behind.  I left some comments below about documentation ... and UX (from my point of view / lets here what other say).

not-my-profile committed:
[simple] improve styling of subcategory table headers in preferences

not-my-profile committed:
[build] /static

not-my-profile committed:
[doc] engine-table: stop sorting by Disabled

It's only the default value for disabled it's not that important,
since users can easily toggle it anytime in the engine preferences.

not-my-profile committed:
[doc] engine tables: show engines in all categories

Previously the documentation grouped the engines by their first
category so e.g. YouTube and Invidious were only shown in the
in the videos section but not in the music section.

This commit fixes this by iterating over searx.engines.categories,
which also has the added benefit that the sections are now in the
same order as the tabs in the user interface.

not-my-profile committed:
[enh] group engines in preference tabs

not-my-profile committed:
[fix] bang syntax for categories with spaces

not-my-profile committed:
[enh] show categories not in any tab category in "Other" preferences tab

Previously we didn't have a good place to put search engines that don't
fit into any of the tab categories. This commit automatically puts
search engines that don't belong to any tab category in an "other"
category, that is only displayed in the user preferences (and not above
search results).

not-my-profile committed:
[enh] move dictionaries, Erowid & IMDb out of general category

The general category is the category that is searched by default.
From a privacy standpoint it doesn't make sense to send all general
queries to specialized search engines that cannot deal with those
queries anyway.

dalf commented on searx/webapp.py:
```diff
@@ -168,8 +172,19 @@
+    gettext(OTHER_CATEGORY),
 )
 
+_powered_by = gettext('powered by %(company)s')
```
No need for this one since there is `_('powered by %(company)s')` in the .html templates.

dalf commented on searx/templates/oscar/preferences.html:
```diff
@@ -298,7 +298,7 @@ <h3>{{ _('Privacy') }}</h3>
                 <!-- Nav tabs -->
                 <ul class="nav nav-tabs nav-justified hide_if_nojs" role="tablist">
-                    {% for categ in all_categories %}
+                    {% for categ in categories_as_tabs %}
```
It should be `categories_as_tabs + [OTHER_CATEGORY]` too otherwise the "Other" tab is missing (Oscar theme)

dalf commented on searx/engines/__init__.py:
```diff
@@ -258,3 +267,28 @@ def load_engines(engine_list):
+        return (powered_by == '', groupname == DEFAULT_GROUP_NAME, groupname.lower())
+
+    def get_group(eng):
+        non_tab_engines = [c for c in eng.categories if c not in settings['categories_as_tabs'] + [OTHER_CATEGORY]]
```
should `non_tab_engines` be `non_tab_categories` ?

dalf commented on searx/engines/__init__.py:
```diff
@@ -258,3 +267,28 @@ def load_engines(engine_list):
+DEFAULT_GROUP_NAME = 'others'
+
+
+def group_engines_in_tab(engines):  # pylint: disable=redefined-outer-name
```
Usually function dedicated to UI are in `searx/webutils.py`

dalf commented on searx/templates/simple/preferences.html:
```diff
@@ -289,12 +289,27 @@ <h2>{{ _('Preferences') }}</h2>
-      {% for search_engine in engines_by_category[categ] %}
+      {% for (group, powered_by), engines in engines_by_category[categ] | group_engines_in_tab %}
+      {% if loop.length > 1 %}
+      <tr><th colspan=9 class=engine-group>
```
it should be `<tr><th colspan="9" class="engine-group">` with quote

not-my-profile committed:
[mod] move group_engines_in_tab to searx.webutils

not-my-profile committed:
[mod] improve variable name

not-my-profile committed:
[enh] display help text for "Other" engines tab in preferences

not-my-profile committed:
[enh] enable English dictionaries by default

These dictionaries are no longer part of the general category,
so they're no longer queried by default -> we can enable them
by default without degrading general query performance.

return42 commented on docs/admin/engines/settings.rst:
```diff
@@ -219,6 +219,26 @@ Communication with search engines.
 ``max_redirects`` :
   30 by default. Maximum redirect before it is an error.
 
+``categories_as_tabs:``
```
nit: is `category_tabs` a better name?

not-my-profile replied:
No.

return42 replied:
I asked because a conjunction (_x_as_y_) in a name of an option is not common.

not-my-profile replied:
`categories_as_tabs` is way clearer than `category_tabs` because it explains that you should list categories there and these categories function as tabs. `category_tabs` does not communicate any of that. "Category tabs" could be something completely different than categories.

return42 commented on searx/webutils.py:
```diff
@@ -134,3 +136,30 @@ def is_flask_run_cmdline():
+DEFAULT_GROUP_NAME = 'others'
+
+
+def group_engines_in_tab(engines):
```
Hard to read/understand, can you please left some hints in a doc-string about the sort order algorithm, the input value and the return value.

not-my-profile replied:
Valid criticism. I refactored the method and added a docstring + type hints.

return42 replied:
Thanks / but does not really describe the sorting algorithm (sort order criteria) ... anyway.

return42 commented on searx/engines/__init__.py:
```diff
@@ -44,6 +44,9 @@
     "tokens": [],
 }
+# set automatically when an engine does not have any tab category
+OTHER_CATEGORY = 'other'
```
Nit: would `NON_CATEGORY = 'uncategorized'` better?

not-my-profile replied:
I don't think so. Firstly "uncategorized" would be completely misleading since engines with categories can be in the "OTHER_CATEGORY". Secondly naming the variable something other than the display text just makes the mental mapping more complicated.

return42 commented on searx/webutils.py:
```diff
@@ -134,3 +136,30 @@ def is_flask_run_cmdline():
     return frames[-2].filename.endswith('flask/cli.py')
+
+
+DEFAULT_GROUP_NAME = 'others'
```
Single-point of definition: should we better use the value from `searx.engines.OTHER_CATEGORY`? --> In webapp.py there is a `from searx.engines import  OTHER_CATEGORY, DEFAULT_GROUP_NAME`.

Do we need two values, one for default one for the "OTHER" tab in the preferences?

not-my-profile replied:
I think for the tab "Other" fits better because general, news, map, music, it, science, social media are all singular.
Whereas the subcategories are almost all plural, e.g. packages, repos & others, apps & others, dictionaries & others.

return42 replied:
But we do not need a "DEFAULT_GROUP_NAME" twice I think.

PLEASE: do not close my remarks as "resolved conversation" .. I will do it .. and I think @dalf, @mrpaulblack should have a chance to read my remarks ...

not-my-profile replied:
>But we do not need a "DEFAULT_GROUP_NAME" twice I think.

OTHER_CATEGORY and DEFAULT_GROUP_NAME fulfill different functions. Using a single variable for different things seems like bad style to me. (Besides that if we want different UI text like I explained we need different variables).

>PLEASE: do not close my remarks as "resolved conversation"

Right, sorry about that. I think it's a bit annoying how GitHub intermixes code review with PR discussion, it makes the comments on the PR hard to follow.

return42 replied:
>  it makes the comments on the PR hard to follow.

Yes, its hard to follow .. I understand.

>> But we do not need a "DEFAULT_GROUP_NAME" twice I think.
> OTHER_CATEGORY and DEFAULT_GROUP_NAME fulfill different functions. 

Ooops sorry, my fail: I thought you have defined DEFAULT_GROUP_NAME twice .. but thats not the case.

return42 commented on searx/settings.yml:
```diff
@@ -1076,10 +1083,12 @@ engines:
     additional_tests:
       rosebud: *test_rosebud
+    about:
+      powered_by: Bing
```
Isn't it the same for the other qwant engines ('qwant images', 'qwant news' ...)?

Are there any reasons why "powered_by" is sometimes set in the settings.yml and sometimes direct in the engine (aka module)?

not-my-profile replied:
>Isn't it the same for the other qwant engines ('qwant images', 'qwant news' ...)?

For web engines it's easy to verify by searching for a site that returns the User-Agent. For images and news it's not so easy to verify.

>Are there any reasons why "powered_by" is sometimes set in the settings.yml and sometimes direct in the engine (aka module)?

Yes it follows the existing convention of defining metadata in the module if possible (e.g. categories are currently defined in the engines when possible).

return42 replied:
> For web engines it's easy to verify by searching for a site that returns the User-Agent. For images and news it's not so easy to verify.

I do not understand this answer .. perhaps I do not understand what "powered by" means .. I thought, the results of this engine (e.g. qwant) are powered by e.g. Bing .. I am wrong? .. what do you mean "User-Agent"?

> Yes it follows the existing convention of defining metadata in the module if possible (e.g. categories are currently defined in the engines when possible).

I asked, because you defined "powered by" of qwant in settings.yml but in bing (itself) and ddg you defined it in the engine module:

![grafik](https://user-images.githubusercontent.com/554536/148056042-660f77d0-c2f4-451e-8887-ce3b041fc96e.png)

not-my-profile replied:
>I thought, the results of this engine (e.g. qwant) are powered by e.g. Bing .. I am wrong? .. what do you mean "User-Agent"?

Search engines like Qwant don't actually disclose in detail which results are powered by what. One way to find out is to find the User-Agent of the result crawler in their search result descriptions.

>I asked, because you defined "powered by" of qwant in settings.yml but in bing (itself) and ddg you defined it in the engine module

Yes because I wanted to avoid categorizing Qwant Video & Images with powered by since I actually don't know that.

return42 replied:
Qwant is powerd by Bing .. but both have different results for the same search term ... 

TBH we do not know which engine is "powered by" which search engine .. AFIAK qwant mix results from bing and from a qwant index.

No matter from which point I look on this "powered by" .. it should never be a sort criteria, since it could never be a serious criteria.

not-my-profile replied:
> it should never be a sort criteria, since it could never be a serious criteria

Of course it can (notice the bingbot):

![image](https://user-images.githubusercontent.com/73739153/148066210-2aecbc4d-441a-4948-9010-b3459eb7740c.png)

return42 replied:
>  (notice the bingbot):

Sorry if I ask a dumb question, but how reliable is the assumption that ddg uses (only) results from a Bing partnership?

not-my-profile replied:
[from their website](https://help.duckduckgo.com/duckduckgo-help-pages/results/sources/):

>We also of course have more traditional links in the search results, which we also source from multiple partners, **though most commonly from Bing**

Please note that I never implied that they "only" use Bing ... but it's their main data source (as can be verified with the User-Agent trick).

not-my-profile committed:
[refactor] make group_engines_in_tab more readable

not-my-profile committed:
[mod] introduce searx.engines.Engine for type hinting

not-my-profile committed:
[fix] stop setting superfluous cookies for "non tab categories"

not-my-profile committed:
[enh] change categories_as_tabs from a list to a dict

The tab icon names are currently hard coded in the templates.
This commit lets us introduce an icon property in the future, e.g:

categories_as_tabs:
  general:
    icon: search-outline

return42 reviewed (approved):
Thanks a lot for you hard work on this topic and thanks for removing "powered by" and opening https://github.com/searxng/searxng/issues/690

@dalf @mrpaulblack I think this PR is now a good starting point for #690 .. lets merge.

I slightly modified (the order of) the topics in #690
